src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -113,11 +134,331 @@ public void Initialize(ChainedHeader consensusTip)
+        }
+
+        /// <summary>
+        /// Handles situation when consensuses tip was changed.
+        /// </summary>
+        /// <param name="newConsensusTip">The new consensus tip.</param>
+        /// <returns>List of peer Ids that violate max reorg rule.</returns>
+        public List<int> ConsensusTipChanged(ChainedHeader newConsensusTip)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(newConsensusTip), newConsensusTip);
+
+            // Switch consensus tip to the new block header.
+            this.AddOrReplacePeerTip(LocalPeerId, newConsensusTip.HashBlock);
+
+            List<int> peerIdsToResync = this.FindPeersToResync(newConsensusTip);
+
+            // Remove block data for the headers that are too far from the consensus tip.
+            this.CleanOldBlockDataFromMemory(newConsensusTip);
+
+            this.logger.LogTrace("(-)");
+            return peerIdsToResync;
+        }
+
+        /// <summary>Checks each peer's tip if it violates max reorg rule.</summary>
+        /// <param name="consensusTip">Consensus tip.</param>
+        /// <returns>List of peers which tips violate max reorg rule.</returns>
+        private List<int> FindPeersToResync(ChainedHeader consensusTip)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(consensusTip), consensusTip);
+
+            var peerIdsToResync = new List<int>();
+            uint maxReorgLength = this.chainState.MaxReorgLength;
+
+            // Find peers with chains that now violate max reorg.
+            if (maxReorgLength != 0)
+            {
+                foreach (KeyValuePair<int, uint256> peerIdToTipHash in this.peerTipsByPeerId)
+                {
+                    ChainedHeader peerTip = this.chainedHeadersByHash[peerIdToTipHash.Value];
+                    int peerId = peerIdToTipHash.Key;
+
+                    ChainedHeader fork = peerTip.FindFork(consensusTip);
+
+                    // Do nothing in case peer's tip is on our consensus chain.
+                    if ((fork != consensusTip) && (fork != peerTip))
+                    {
+                        int reorgLength = consensusTip.Height - fork.Height;
+
+                        if (reorgLength > maxReorgLength)
+                        {
+                            peerIdsToResync.Add(peerId);
+                            this.logger.LogTrace("Peer with Id {0} claims a chain that violates max reorg, its tip is '{1}'.", peerId, peerTip);
+                        }
+                    }
+                }
+            }
+
+            this.logger.LogTrace("(-):*.{0}={1}", nameof(peerIdsToResync.Count), peerIdsToResync.Count);
+            return peerIdsToResync;
+        }
+
+        /// <summary>Cleans the block data for chained headers that are old. This data will still exist in the block store if it is enabled.</summary>
+        /// <param name="consensusTip">Consensus tip.</param>
+        private void CleanOldBlockDataFromMemory(ChainedHeader consensusTip)
```

Comment:
I am starting to reconsider `ChainedHeader.Block` this should be available in its own Interface.